# FIT5196 Assessment 1: 
# Task1: Parsing Raw Text Files
#### Student Name: Prashantkumar P Kulkarni
#### Student ID: 29470811

Date:01-09-2018

Environment: Python 3.6 and Jupyter notebook
Libraries used: 
* JSON (for converting to JSON format)
* re (for regular expression, included in Anaconda Python 2.7) 

### Importing Libraries

In [6]:
import re
import json

### Reading the file and splitting the file into list of resumes
 1) Open the dataset file given, in Read mode
 2) Split the resume into the list of individual resumes
 3) Remove the last element as it is null

In [20]:
file = open('29470811.dat','r')
file_read = file.read()
file.close()
resume_list = file_read.split('------------------------------')
resume_list.remove(resume_list[30735])

### Identifying the pattern of the keys 
From the observation of some individual resumes, we find that keys start at the begining of the line and hence precede by \n.         
1) UpperCase keys: The pattern of uppercase letters are                                                               
 a) Single Uppercase Keys: These are the keys which have a single word and a colon followed (Eg: 'KEY:')         
 b) Double Uppercase Keys: These are the keys which have two words seperated by either special characters('-','_' )or 'Space' (Eg: key_key)
2) LowerCase with special characters: These keys are the single word keys preceeded by ('-','_' ) (Eg: '_key)
3) LowerCase keys: Unique LowerCase keys consisting of single word followed colon('key:') 

 The uppercase keys and Lower Case keys with special characters are identifiable, but LowerCase keys are not, so we will analyse and filter the only valid keys out of it.

In [26]:
small_keys = []
for each in resume_list:
    small_keys= small_keys + (re.findall(r'\n[a-z]*:',each))
print(len(set(small_keys)))

175


Out of 175 keys after analysing all these keys, we conclude that only \nsalary:, \nresponsibilities:, \ndeadline:, \nprocedures:,\nqualifications:, \nremuneration:,\ntitle: are the only valid keys.

#### Converting all keys into a proper format
Having all the keys in one proper format is important to fetch all the keys easily and hence we convert all the keys to a proper format





In [30]:
updated_resume_list = []
for each in resume_list:

    lower_keys = re.findall(r'\n[a-z]*[-_/]+[a-z]+:|\nsalary:|\nresponsibilities:|\ndeadline:|\nprocedures:|\nqualifications:|\nremuneration:|\ntitle:',each)
    upper_keys = re.findall(r'\n[A-Z]+ [A-Z]+:',each)
    
    for value in lower_keys:
        if (value):
            each=each.replace(value,value.upper())
        
    for value in upper_keys:
        if (value):
            each = each.replace(value,value.replace(' ','_'))
    updated_resume_list.append(each)

#### Now the updated list will have the keys which are all in uppercase(Either with or without '-','_')

### Creating a list of unique keys with similar pattern
First will create a list of all keys with similar pattern(Keys having same meaning)
Convert all the keys to a general key for all the resumes and add it to a new list

In [43]:
resume_list = []
for each in updated_resume_list:
    
    reg_about = re.findall(r'\nABOUT.*|\n.*INFO.*:',each)
    reg_location = re.findall(r'\n.*LOC.*:',each)
    reg_deadline = re.findall(r'\nDEAD.*:|\nAPP.*:',each)
    reg_procedure = re.findall(r'\n.*PROC.*:',each)
    reg_start = re.findall(r'\n.*DA.*:',each)
    reg_description = re.findall(r'\n.*DESC.*:',each)
    reg_salary = re.findall(r'\n.*SAL.*:|\n.*REMU.*:',each)
    reg_responsibility = re.findall(r'\n.*RESP.*:',each)
    reg_title = re.findall(r'\nJOB_T.*:|\n_TTL:|\n.*TITLE.*:',each)
    reg_qualification = re.findall(r'\n.*QUAL.*:',each)
    reg_responsibility = re.findall(r'\nRESP.*:',each)
    
    for value in reg_about:
        if(value):
            each = each.replace(value,'\nABOUT_COMPANY:')
    for value in reg_location:
        if(value):
            each = each.replace(value,'\nJOB_LOCATION:')
    for value in reg_deadline:
        if(value):
            each = each.replace(value,'\nAPPLICATION_DEADLINE:')
    for value in reg_procedure:
        if(value):
            each = each.replace(value,'\nPROCEDURE:')
    for value in reg_start:
        if(value):
            each = each.replace(value,'\nSTART_DATE:')
    for value in reg_description:
        if(value):
            each = each.replace(value,'\nJOB_DESCRIPTION:')
    for value in reg_salary:
        if(value):
            each = each.replace(value,'\nSALARY:')
            
    for value in reg_responsibility:
        if(value):
            each = each.replace(value,'\nJOB_RESPONSIBILITIES:')
    for value in reg_title:
        if(value):
            each = each.replace(value,'\nJOB_TITLE:')
    for value in reg_qualification:
        if(value):
            each = each.replace(value,'\nJOB_QUALIFICATIONS:')
            
    resume_list.append(each)

### Fetching the keys and values using regular expression

All the keys and their respective values will be fetched and stored in a dictionary

In [202]:
dictionary =[]
for each in resume_list:
    
    reg = re.finditer(r'([A-Z]*[' ']*[-_/]*[A-Z]+:/*\n*)([\s\S]*?)(?=[A-Z]*[ -_/]*[A-Z]+[:/]+|$)',each, re.DOTALL)
    ID = ""
    title = ""
    salary = "" 
    start = "" 
    about = "" 
    qualification = []
    procedure = "" 
    responsibility = []
    deadline = "" 
    remuneration = "" 
    description =[]
    location = "" 
    value_title = "N/A"
    value_about = "N/A"
    value_start = "N/A"
    value_qualification = ["N/A"]
    value_salary = "N/A"
    value_procedure = "N/A"
    value_location = "N/A"
    value_deadline = "N/A"
    value_responsibility = ["N/A"]
    value_description = ["N/A"]
    value_remuneration = "N/A"
    listing = []
    
    for each in reg:
        
            if (each.group(1)=='ID:'):
                value_id = each.group(2).strip()
                value_id = value_id.replace('\n','')
                ID = ID + str(value_id)
                
            elif each.group(1) in ('JOB_TITLE:','JOB_TITLE:\n'):
                if(each.group(2)):
                    value_title = each.group(2).replace('\n','')
                    value_title = value_title.strip()
                    
            elif each.group(1) in ('ABOUT_COMPANY:','ABOUT_COMPANY:\n'):
                if(each.group(2)):
                    value_about = each.group(2).replace('\n','')
                    value_about = value_about.strip()
                    
            elif each.group(1) in ('START_DATE:','START_DATE:\n'):
                if(each.group(2)):
                    value_start = each.group(2).replace('\n','')
                    value_start = value_start.strip()
                    
            elif each.group(1)in('SALARY:','SALARY:\n'):
                if(each.group(2)):
                    value_salary = each.group(2).replace('\n','')
                    value_salary = value_salary.strip()
                    
            elif each.group(1) in ('PROCEDURE:','PROCEDURE:\n'):
                if(each.group(2)):
                    value_procedure = each.group(2).replace('\n','')
                    value_procedure = value_procedure.strip()
                    
            elif each.group(1)in('JOB_LOCATION:','JOB_LOCATION:\n'):
                if(each.group(2)):
                    value_location = each.group(2).replace('\n','')
                    value_location = value_location.strip()
                    
            elif each.group(1)in('APPLICATION_DEADLINE:','APPLICATION_DEADLINE:\n'):
                if(each.group(2)):
                    value_deadline = each.group(2).replace('\n','')
                    value_deadline = value_deadline.strip()
                    
                        
            elif each.group(1)in('JOB_DESCRIPTION:','JOB_DESCRIPTION:\n'):
                if(each.group(2)):
                    value_desc = each.group(2).strip()
                    #value_desc = value_desc.replace('. ','.')
                    value_desc = value_desc.replace('\n',' ')
                    value_desc = value_desc.split('. ')
                    value_description = []
                    for each in value_desc:
                        if(each!=each.upper()):
                            each = each.strip()
                            #each = each.replace('\','')
                            value_description.append(each)
                    
            elif each.group(1)in('JOB_RESPONSIBILITIES:','JOB_RESPONSIBILITIES:\n'):
                if(each.group(2)):
                    value_response = each.group(2).strip()
                    value_response = value_response.split('\n-')
                    value_responsibility = []
                    for each in value_response:
                        if(each!=each.upper()):
                            each = each.strip()
                            each = each.replace('\n',' ')
                            each = each.replace('- ','')
                            each = each.replace(';','')
                            each = each.replace('.','')
                            value_responsibility.append(each)
                        
        
            elif each.group(1)in('JOB_QUALIFICATIONS:','JOB_QUALIFICATIONS:\n'):
                if(each.group(2)):
                    value_qual = each.group(2).strip()
                    value_qual = value_qual.split('\n-')
                    value_qualification = []
                    for each in value_qual:
                        if(each!=each.upper()):
                            each = each.strip()
                            each = each.replace('\n',' ')
                            each = each.replace(';','')
                            each = each.replace('.','')
                            value_qualification.append(each)
            
    title = title + str(value_title)
    salary = salary+str(value_salary)
    start=start+str(value_start)
    about=about+str(value_about)
    qualification=qualification+value_qualification
    procedure=procedure+str(value_procedure)
    responsibility=responsibility+value_responsibility
    deadline=deadline+str(value_deadline)
    #remuneration=remuneration+str(value_remuneration)
    description=description+(value_description)
    location=location+str(value_location)
    
    dictionary_listing = {'_id':ID,'title':title,'location':location,'job_descriptions':{'description':description},'job_responsibilities':{'responsibility':responsibility},
             'required_qualifications':{"qualification":qualification},'salary':salary,'application_procedure':procedure,'start_date':start,'application_deadline':deadline,
             'about_company':about}
    
    listing.append(dictionary_listing)
    
    dictionary.append({"listing":listing})
    


json_dictionary = {'listings' : dictionary1}

### Using JSON's json.dumps we convert our final dictionary to JSON file

In [255]:
with open('29470811.json', 'w') as file:
    file.write(json.dumps(json_dictionary, indent = 4))